Rewrite Part 2

iliis opened the pull request:
So, Schluss für heute. Das wichtigste ist da und läuft, wenn auch noch nicht ganz korrekt... Basically sind wir wieder gleich weit wie vorher, dafür mit viel schönerem Code.

Die Änderungen im Schnellüberblick:
- Es werden jetzt überall Spaltenvektoren verwendet.
- Die Koordinatensysteme entsprechen jetzt denen in https://github.com/iliis/BA/raw/rewrite/matlab/coordinate_systems.pdf
  Die umgekehrte V-Achse der Bildkoordinaten ist irgendwie echt dämlich. Man muss jetzt aufpassen, dass die Itensitäts- und Tiefen-Bilder aufm Kopf übergeben werden, bzw. immer intensity_sample() verwenden :/
- vollautomatische Jacobians sind implementiert: die Funktionen rufen sich selber symbolisch auf und leiten sich dann selber ab. Sehr elegant und saulangsam.
- Ein paar weitere Tests

Was als nächstes zu tun ist:
- Wie rum wollt ihr eure Euler Winkel? ZYZ wie hier http://www.easyspin.org/documentation/eulerangles.html? ZYX wie das Matlab per default macht?
- Wie rum soll die Tranformation gehen?
  Momentan werden Punkte mit R*P+T verarbeitet. Das ist simpel, entspricht aber dem inversen der eigentlichen Kamerabewegung (er werden Punkte und nicht Koordinatensysteme/Kameras verschoben). Ich fänds eigentlich eleganter, wenn bei der Odometrie direkt interpretierbare Zahlen rauskommen, aber wie ihr wollt.
- Sobald obere Punkte geklärt sind Test für die entsprechenden Funktionen schreiben (hauptsächlich camera_transform() und dann camera_warp())
- Tests für die Jacobians. Evtl. mal ein oder zwei Fälle von Hand durchrechnen und sicher numerisch ableiten.

Falls du mit dem Code rumspielen willst:
- setup.m ausführen (da muss das cd() wohl erst rausgenommen werden, evtl. vergessene Pfade hinzufügen ;))
- Testscripte sind in test_scripts: test_simple_warping.m plottet die gewarpten Punkte und test_minimization_gui.m lässt die verschiedenen Minimierungsverfahren laufen. Gradient Descent läuft auf den ersten Blick in die richtige Richtung, aber auch nur wenn man den Gradienten invertiert...
- unittests können mit run_unittests.m laufengelassen werden (die müssten alle durchlaufen)

Have fun...

iliis committed:
Changed some details after last discussion.

iliis committed:
Cleanup. Add sceletons for core functions.

iliis committed:
Implement core functions.

Basic warping seems to be working.

TODO:
- derive/implement Jacobis
- write more tests (especially hand picked samples) for:
    - camera_transformation
    - camera_project
- write helper function to read in complete test scene (with ground
  truth etc.)
- clean up helper functions (remove unnecessary ones, document and test others)

iliis committed:
Add helpful diagram to show coordinate system orientations.

iliis committed:
Fix camera_transform: Use conventional R*X+T.

iliis committed:
Update coordinate system chart to reflect JR's wishes.

WARNING: code was NOT updated! Chart and code do NOT correspond!

iliis committed:
Use column vectors instead of row vectors.

Add tests for image_to_list/list_to_image.
Add script to run all testscripts.

iliis committed:
Update intensity_sample() to use new coordinate systems.

[1,1] = bottom left
[W,H] = top right

iliis committed:
Switched to new coordinate system (see coordinate_systems.pdf)

TODO: write [more] tests for all core functions

iliis committed:
Add sceletons for transformation tests and utilities.

TODO: Fix Euler angles. Fix direction of transformation (move Camera or
move points).

iliis committed:
Implemented Jacobians. No tests, apart from intensity_sample().

iliis committed:
Implement camera_warp(): combines all the core functions.

Unfortunately, it's not very clean (flipped V-axis, no tests)

iliis committed:
Play around with Jacobians.

Not much luck: Real slow and not very correct :(

PascalGohl commented:
sieht auf den ersten blick schon sehr gut aus. bei mir laufen die test aber nicht alle erfolgreich. evt hab ich gerade pech/glück mit den random zahlen. ich empfehle daher in unitest einen fixen seed für die zufallszahlen zu verwenden und dafür aber mehrere durchläufe testen.

PascalGohl commented:
Hier Antworten zu deinen Fragen:
- Ich verwenden meistens ZYX euler winkel, das ist aber leider nicht einheitlich geregelt bei uns im Lab. (Jörn?)
- odometrie output sollte schlussentlich schon die kamera position sein, aber code einfach halten ist auch wichtig. ich schau mir das mal im code an. evt hab ich deine frage falsch verstanden.

PascalGohl commented:
so viel ich bis jetzt in deinem code verstanden haben würde ich die transformation so lassen. vereinfacht die mathe und dadurch das fehlerrisiko.

PascalGohl commented:
könntest du eine funktion schreiben, in der einmal die ganze pipeline durchgegangen wird? also bilder rein, odometrie raus. einfach, damit ich einen überblick kriege über deine funktionen und wie du sie verwendest.

PascalGohl commented:
Ich finde du bist auf einem gute Weg und hast schon fast alles wichtige implementiert. Wann denkst du kannst du die TODOs mit den tranformationen abarbeiten?

PascalGohl commented on matlab/core/camera_warp.m:
```diff
@@ -0,0 +1,107 @@
+if calc_jacobian
+    J_T = J_T(:,:,valid_points);
+    J_P = J_P(:,:,valid_points);
+    [intensities_current, J_I] = camera_intensity_sample(points_current_camera, image_current);
```
Ich glaub die Jacobian sollte vom keyframe berechnet werden. Scheint ein typo in Sammy's paper eq. 11 zu sein. Im text spricht er von keyframe.

iliis replied:
Ich glaube das ist schon korrekt so:
"J_I is the image derivative of the warped image"
"By assuming that the warped image is sufficiently close to the keyframe image, we can approximate the derivative ofthe warped image by the keyframe image derivative."

PascalGohl replied:
Da hast du recht, wir habens damals nur anders rum implementiert. das hat mich verwirrt. So mussten wir nur von den keyframes tiefenbilder berechen.

PascalGohl commented on matlab/core/camera_warp.m:
```diff
@@ -0,0 +1,107 @@
+    assert(N == size(J_P,3));
+    assert(N == size(J_I,3));
+    J_warp = zeros(N,numel(T));
+    % this seems to be one of the fastest way of doing N matrix multiplications:
```
strange, this looks quite slow...

rehderj replied:
Da bin ich mir auch unsicher, ob dies der schnellste Weg ist. Unter Umständen könntest Du die Schleife durch `reshape`-, `.*`\- und `sum`-Operationen umgehen und Deinen Code dadurch deutlich beschleuningen.

iliis replied:
Ists die mehrdimensionale Matrixmultiplikation die langsam ist? Bei mir ist vorallem die symbolische Ableitung ewigs langsam, aber ich müssts mal profilen.

rehderj replied:
Auch möglich. Dann berechne die doch einfach einmalig symbolisch und kopier sie darein. Das wolltest Du aber wahrscheinlich sowieso, sobald der Code einmal auf Richtigkeit geprüft ist.

rehderj commented:
Erstmal eine Reihe von High-Level-Kommentaren:

Ich gebe Dir recht, dass das Invertieren der V-Achse recht dämlich ist, und wenn Du Slide 10 des von mir zuvor verlinkten Dokumnets anschaust, wirst Du sehen, dass es auch nicht konform ist mit der Konvention in der CV-Literatur. Dazu wollte ich gleich aber noch individuelle Zeilen kommentieren.

Ich argumentierte, dass R und T ebenfalls die Kamerabewegung beschreiben und auch unmittelbar interpretierbar sind.

Euler-Winkel finden eigentlich grundsätzlich wenig Verwendung im Lab. Entsprechend verbleiben wir mit Pascals Vorschlag, denke ich.

rehderj commented on matlab/test_scripts/test_simple_warping.m:
```diff
@@ -0,0 +1,22 @@
+% loads an image pair and warps the keyframe onto the current frame
+
+image_path = 'input/testscene1';
+T = [2 0 -4 0 0 0]'; % TODO: read this from camera_trajectory.csv
```
Ha! Diese Zeile scheint nicht konsistent mit [jener](https://github.com/iliis/BA/blob/rewrite/matlab/input/testscene1/camera_trajectory_relative.csv#L3) zu sein, und wenn die CSV-Datei korrekte Transformationen enthält, sollte einem das ein Zeichen sein, dass etwas im Argen liegt.

iliis replied:
Nope, Blender braucht nochmals andere Konventionen und das export-script hab ich noch nicht angepasst solang nocht nicht alle Konventionen festgelegt sind.

rehderj commented on matlab/core/camera_warp.m:
```diff
@@ -0,0 +1,107 @@
+points_keyframe_camera = image_to_list(cat(3,X,Y));
+
+% attention: inverted V-Axis -> flip depths!
+points_world = camera_project_inverse(points_keyframe_camera, image_to_list(flipud(depths_keyframe)), intrinsics);
```
Warum braucht es das `flipud` hier?

iliis replied:
Wegen der invertierten V-Achse. Ist definitiv der falsche Ort, aber ich war da zu müde, das noch anzupassen. Im Prinzip müsste man dann auch die intensity_sample() Funktion verwenden, die dann unnötigerweise interp2 benutzt...

rehderj commented on matlab/core/camera_project_inverse.m:
```diff
@@ -0,0 +1,64 @@
+% [x,y]' = ([u v]' - principal_point) * depth / focal_length
+%    z   = depth
+points_world = [ ...
+    -(points_camera(1,:) - intrinsics.principal_point(1)) .* depths ./ intrinsics.focal_length; % -X = +U !
```
Dieses Minus scheint mir nicht richtig.

rehderj commented on matlab/core/camera_project.m:
```diff
@@ -0,0 +1,63 @@
+
+% X-Axis is mirrored relative to U-Axis in camera plane
+% --> U = -X / Z * focal + principal
+points_world(1,:) = -points_world(1,:);
```
Bitte ebenfalls entfernen.

iliis replied:
AAArgh. Doch, deswegen hab ich ja etwa drei mal nachgefragt, ob du wirklich sicher bist, dass du die Konventionen so willst. Auf Slide 10 geht die U Achse (Xcam/Ximg) klar nach Rechts, während die global X-Achse nach links zeigt (Slide 6). Ich hätts ja auch liebend gern anders :P

rehderj commented on matlab/core/camera_intensity_sample.m:
```diff
@@ -0,0 +1,71 @@
+assert(all(all(points_camera(2,:) <= H)));
+
+% [1,1] is bottom left in image coordinates -> invert V-axis!
+image = flipud(image);
```
Dies scheint mir ebenfalls nicht konform mit den üblichen Konventionen.

iliis replied:
Siehe Slide 10 in https://github.com/iliis/BA/raw/rewrite/matlab/04-E-camera.pdf
Da geht die V-Achse (Yimg/Ycam) doch von unten nach oben?

rehderj replied:
Die punch line sollte sein, dass Kamera- und Bildkoordinatensystem nicht zueinander rotiert sind.

iliis commented:
> Wann denkst du kannst du die TODOs mit den tranformationen abarbeiten?

Sobald mir klar ist, wie all die Konventionen aussehen. Also wie R und T zu interpretieren sind, wie die Euler-Winkel definiert sein sollen, etc.

> Ich argumentierte, dass R und T ebenfalls die Kamerabewegung beschreiben und auch unmittelbar interpretierbar sind.

Richtig, aber sie beschreiben halt die inverse Bewegung. Also von current frame zu keyframe. Ich schliesse mich da Pascal an, dass X2 = RX+T schöner ist als X2 = R'X+R'T

> Euler-Winkel finden eigentlich grundsätzlich wenig Verwendung im Lab. Entsprechend verbleiben wir mit Pascals Vorschlag, denke ich.

Exzellent, aso ZYX. Dh. die Rotationsmatrix R = ZYX sieht dann so aus:

![3bcae89af1478010ee722e680af0afa9](https://cloud.githubusercontent.com/assets/933300/7814408/46635f4a-03c2-11e5-8190-471d3659c312.png)

Wobei Z =

```
cos(a) -sin(a) 0
sin(a)  cos(a) 0
0       0      1
```

Y =

```
 cos(b) 0 sin(b)
 0      1     0
-sin(b) 0 cos(b)
```

X =

```
1     0       0
0 cos(c) -sin(c)
0 sin(c)  cos(c)
```

iliis committed:
Fix coordinate systems for the 3rd (and hopefully last!) time.

iliis committed:
Optimize symbolic Jacobian evaluation: Huge speedup.

Gradient Descent actually doesn't seem *so* bad, but it converges very
slowly in Z direction (and has to be constrained to translation only).

iliis committed:
Add weights to Gauss Newton. Clean up code.

Load camera intrinsics from CSV, add Scene class, add script for
plotting the cost surface.

iliis committed:
Reverse warping: warp current image on to keyframe.

iliis committed:
More cleanup, fix Blender script.

iliis committed:
There's still something slightly wrong with the Blender script.

Differences between Euler angles is almost correct, but blender uses a
rotated frame (180 degrees around X-Axis)

iliis committed:
Add longer trajectory.

iliis committed:
Reorder rotation sub-matrixes. Blender ground truth matches Matlab.

Altough the rotation matrix isn't really ZYX anymore...

iliis committed:
Add support for longer trajectories.

iliis committed:
Initial commit for C++ code.

iliis committed:
Add simple Vim keybinding to compile and run code

iliis committed:
Fix cost surface renderer (use SceneStep).

iliis committed:
Change build system to CMake. Test Eigen/SFML.

iliis committed:
WIP on C++ implementation. Change depth images to 8bit PNGs.

SFML/stb_image cannot handle more than 8bit per channel :( But real
depth data will be <8bit anyway.

iliis committed:
Add a few interesting plots of the cost surface.

iliis committed:
Add CameraIntrinsics class to C++ codebase.

iliis committed:
Add Transformation class to C++ codebase.

iliis committed:
Basic warping implemented in C++.

No Jacobians.
Code is a bit ugly.
Not optimized at all (needs about 0.8s to warp an 256x128 image, Matlab
uses about 1.4s)

iliis committed:
Plot nice error plots in C++. 6ms vs 2h :P

iliis committed:
Add nice colormaps :)

iliis committed:
Add fancy progress bar for console.

iliis committed:
Basic Jacobian calculations in C++. Not correct!

Various other small things.

iliis committed:
Add short Readme about C++ code (how to build).

iliis committed:
Add another errorplot.

iliis committed:
Fix output_frame_pos.py: Revert to correct version.

Was probably overwritten with older version by Blender.
Also, opening another file with scripts still registered (?) seems to
cause issues.

iliis committed:
...
